### ISSUE TITLE
history dataset association ID is out of range for type integer.

### ISSUE THREAD

**Issue by spficklin** [galaxyproject/galaxy#6299]
We are using blend4php (http://github.com/galaxyproject/blend4php) to invoke a Galaxy workflow. It works fine for all workflows except one.   The error we are receiving is an API error from Galaxy itself.  The problem arises when setting the 'inputs' argument that gets POST'ed to the Galaxy API.  Here's an example of that array that gets posted (It's in PHP + JSON encoded format so sorry for the weirdness  of it but you can see what parts are there):
```
Array
(
    [inputs] => {"0":{"id":"1265663246886131","src":"hda"}}
    [dataset_map] => {"0":{"id":"1265663246886131","src":"hda"}}
    [history_id] => 076f9819daba3c1c
    [parameters] => {"0":{"vcf_input":"1265663246886131"}}
)
```
When setting the 'id' for the first input, the blend4php code always uses the 'id' for the dataset.  Here's what we get back for the history contents from the Galaxy API (URL used https://{our.galaxy.server}/api/histories/076f9819daba3c1c/contents):
```
[
    {
        "create_time": "2018-06-07T01:52:27.065608", 
        "dataset_id": "d0ff981e009477e4", 
        "deleted": false, 
        "extension": "vcf", 
        "hid": 1, 
        "history_content_type": "dataset", 
        "history_id": "076f9819daba3c1c", 
        "id": "1265663246886131", 
        "name": "OA.vcf", 
        "purged": false, 
        "state": "ok", 
        "tags": [], 
        "type": "file", 
        "type_id": "dataset-1265663246886131", 
        "update_time": "2018-06-07T01:52:49.661627", 
        "url": "/api/histories/076f9819daba3c1c/contents/1265663246886131", 
        "visible": true
    }
]
```
You can see that the 'id' used in the array that is sent via POST for invoking a workflow is the same as in the history contents.  And the code that generates that array for invoking a workflow works the same for every workflow we want to invoke.  However, for this particular dataset, when we try to use it in the workflow we get this error message:

```
Uncaught exception in exposed API method: Traceback (most recent call last): 
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/web/framework/decorators.py", line 281, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/webapps/galaxy/api/workflows.py", line 641, in invoke
    request_params=work_request_params
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/workflow/run.py", line 108, in queue_invoke
    modules.populate_module_and_state(trans, workflow, workflow_run_config.param_map,
allow_tool_state_corrections=workflow_run_config.allow_tool_state_corrections)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/workflow/modules.py", line 1117, in populate_module_and_state
    step_errors = module_injector.inject(step, step_args=step_args)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/workflow/modules.py", line 1096, in inject
    state, step_errors = module.compute_runtime_state(self.trans, step_args)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/workflow/modules.py", line 772, in compute_runtime_state
    state, step_errors = super(ToolModule, self).compute_runtime_state(trans, step_updates)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/workflow/modules.py", line 203, in compute_runtime_state
    visit_input_values(self.get_runtime_inputs(), state.inputs, update_value, no_replacement_value=NO_REPLACEMENT)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/tools/parameters/__init__.py", line 112, in visit_input_values
    callback_helper(input, input_values, name_prefix, label_prefix, parent_prefix=parent_prefix, context=context)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/tools/parameters/__init__.py", line 77, in callback_helper
    new_value = callback(**args)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/workflow/modules.py", line 197, in update_value
    value, error = check_param(trans, input, step_updates.get(prefixed_name), context)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/tools/parameters/__init__.py", line 132, in check_param
    value = param.from_json(value, trans, param_values)
  File "/main/sites/galaxy/galaxy/htdocs/lib/galaxy/tools/parameters/basic.py", line 1666, in from_json
    rval = trans.sa_session.query(trans.app.model.HistoryDatasetAssociation).get(value)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 831, in
get
    return self._get_impl(ident, loading.load_on_ident)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 864, in
_get_impl
    return fallback_fn(self, key)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/loading.py", line 219,
in load_on_ident
    return q.one()
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2718, in
one
    ret = list(self)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2761, in
__iter__
    return self._execute_and_instances(context)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2776, in
_execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 914,
in execute
    return meth(self, multiparams, params)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 323,
in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1010,
in _execute_clauseelement
    compiled_sql, distilled_params
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1146,
in _execute_context
    context)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1341,
in _handle_dbapi_exception
    exc_info
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 202,
in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1139,
in _execute_context
    context)
  File "/main/sites/galaxy/galaxy/htdocs/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line
450, in do_execute
    cursor.execute(statement, parameters)
DataError: (psycopg2.DataError) value "1265663246886131" is out of range for type integer
LINE 3: WHERE history_dataset_association.id = '1265663246886131'
                                               ^
 [SQL: 'SELECT history_dataset_association.id AS history_dataset_association_id, history_dataset_association.history_id
AS history_dataset_association_history_id, history_dataset_association.dataset_id AS
history_dataset_association_dataset_id, history_dataset_association.create_time AS
history_dataset_association_create_time, history_dataset_association.update_time AS
history_dataset_association_update_time, history_dataset_association.state AS history_dataset_association_state,
history_dataset_association.copied_from_history_dataset_association_id AS
history_dataset_association_copied_from_history_dataset_a_1,
history_dataset_association.copied_from_library_dataset_dataset_association_id AS
history_dataset_association_copied_from_library_dataset_d_2, history_dataset_association.name AS
history_dataset_association_name, history_dataset_association.info AS history_dataset_association_info,
history_dataset_association.blurb AS history_dataset_association_blurb, history_dataset_association.peek AS
history_dataset_association_peek, history_dataset_association.tool_version AS history_dataset_association_tool_version,
history_dataset_association.extension AS history_dataset_association_extension, history_dataset_association.parent_id AS
history_dataset_association_parent_id, history_dataset_association.designation AS
history_dataset_association_designation, history_dataset_association.deleted AS history_dataset_association_deleted,
history_dataset_association.visible AS history_dataset_association_visible,
history_dataset_association.extended_metadata_id AS history_dataset_association_extended_metadata_id,
history_dataset_association.hid AS history_dataset_association_hid, history_dataset_association.purged AS
history_dataset_association_purged, history_dataset_association.hidden_beneath_collection_instance_id AS
history_dataset_association_hidden_beneath_collection_ins_3, dataset_1.id AS dataset_1_id, dataset_1.create_time AS
dataset_1_create_time, dataset_1.update_time AS dataset_1_update_time, dataset_1.state AS dataset_1_state,
dataset_1.deleted AS dataset_1_deleted, dataset_1.purged AS dataset_1_purged, dataset_1.purgable AS dataset_1_purgable,
dataset_1.object_store_id AS dataset_1_object_store_id, dataset_1.external_filename AS dataset_1_external_filename,
dataset_1._extra_files_path AS dataset_1__extra_files_path, dataset_1.file_size AS dataset_1_file_size,
dataset_1.total_size AS dataset_1_total_size, dataset_1.uuid AS dataset_1_uuid \nFROM history_dataset_association LEFT
OUTER JOIN dataset AS dataset_1 ON dataset_1.id = history_dataset_association.dataset_id \nWHERE
history_dataset_association.id = %(param_1)s'] [parameters: {'param_1': u'1265663246886131'}]

```
Notice from above this message:
```
DataError: (psycopg2.DataError) value "1265663246886131" is out of range for type integer
LINE 3: WHERE history_dataset_association.id = '1265663246886131'

```
I am not sure how to debug this issue.  Any thoughts on what might be going wrong?  We're using Galaxy Server v17.09.

**Comment by NicHerndon** [comment]
```
--- a/lib/galaxy/tools/parameters/basic.py
+++ b/lib/galaxy/tools/parameters/basic.py
@@ -1656,6 +1656,8 @@ class DataToolParameter(BaseDataToolParameter):
                 elif isinstance(single_value, trans.app.model.HistoryDatasetAssociation):
                     rval.append(single_value)
                 else:
+                    if single_value.isdigit() and len(str(single_value)) == 16:
+                        single_value = trans.security.decode_id(single_value)
                     rval.append(trans.sa_session.query(trans.app.model.HistoryDatasetAssociation).get(single_value))
             if found_hdca:
                 for val in rval:
@@ -1681,6 +1683,7 @@ class DataToolParameter(BaseDataToolParameter):
         elif isinstance(value, trans.app.model.HistoryDatasetCollectionAssociation):
             rval = value
         else:
+            value = trans.security.decode_id(value)
             rval = trans.sa_session.query(trans.app.model.HistoryDatasetAssociation).get(value)
         values = util.listify(rval)
         dataset_matcher_factory = get_dataset_matcher_factory(trans)
```

seems to have been necessary

**Comment by hexylena** [comment]
I think this was fixed in #6433
